### PR TITLE
[EMB-171] Handle toMany relationships for create/update in v2 api

### PR DESCRIPTION
**Pull request by brianjgeiger**
## Purpose

In EmberOSF, we have some terrible hacks in to handle updating relationships that go to many relationships. In particular, the affiliated_institutions relationship on nodes. The reason is that the preferred method of updating that information, by including it in the relationship field of the POST or PATCH request, only worked with 1 to 1 relationships, such as Preprint->Node or Node->License. This generalizes everything to allow that to work with toMany relationships, and in particular adds support for Node->affiliated_institutions.

## Changes

1. Update parsers to handle lists
2. Update affiliated_institutions field to be writeable, not be required, and have "many"
3. Update base serializer to handle where everything broke when "many" was activated (mostly related to not realizing that fields also have child_relations that need to be checked)
4. Add parser classes to node list view so creation worked
5. Update tests to handle relationships that were lists

## QA Notes

If preprints creation/updating still works, then this should all be fine. We aren't using the multiple feature until the Dashboard is in, and it's mostly making sure the rest of the API didn't break.  Any runscope tests that create or patch with relationships (preprints or node licenses) would be good to run as well, though we have pretty thorough test coverage on this.

## Side Effects

Potentially. The most suspicious change was overriding the `to_internal_value` method on the Relationship field, which was weird that it hadn't been necessary before but is now. I suspect it's somehow related to having that field be accessed indirectly with `child_relations`.

## Ticket

https://openscience.atlassian.net/browse/EMB-171

### CODE REVIEW

File: api/nodes/serializers.py [hunk 1]
```diff
@@ -35,6 +35,45 @@
...
+def get_institutions_to_add_remove(institutions, new_institutions):
+    diff = relationship_diff(
+        current_items={inst._id: inst for inst in institutions.all()},
+        new_items={inst['_id']: inst for inst in new_institutions}
+    )
+
+    insts_to_add = []
+    for inst_id in diff['add']:
+        inst = Institution.load(inst_id)
+        if not inst:
```
Comment (sloria) [inline api/nodes/serializers.py]: Minor: You can do `Institution.objects.filter(_id=inst_id).exists()` here.

File: api/nodes/serializers.py [hunk 2]
```diff
@@ -35,6 +35,45 @@
...
+        new_institutions=new_institutions
+    )
+
+    if post and not len(add):
+        raise RelationshipPostMakesNoChanges
+
+    if not post:
+        for inst in remove:
+            if not user.is_affiliated_with_institution(inst) and not node.has_permission(user, 'admin'):
+                raise exceptions.PermissionDenied(
```
Comment (sloria) [inline api/nodes/serializers.py]: This looks like something that should go on the view's permission class.

File: api/nodes/serializers.py [hunk 3]
```diff
@@ -1016,21 +1068,6 @@ def get_related_url(self, obj):
...
-        diff = relationship_diff(
-            current_items={inst._id: inst for inst in institutions.all()},
-            new_items={inst['_id']: inst for inst in new_institutions}
-        )
-
-        insts_to_add = []
-        for inst_id in diff['add']:
-            inst = Institution.load(inst_id)
-            if not inst:
-                raise exceptions.NotFound(detail='Institution with id "{}" was not found'.format(inst_id))
```
Comment (brianjgeiger) [inline api/nodes/serializers.py]: @sloria Just for the record, those items weren't mine, I was just moving them from here.

File: api/nodes/serializers.py [hunk 4]
```diff
@@ -1040,22 +1077,7 @@ def make_instance_obj(self, obj):
...
         node = instance['self']
         user = self.context['request'].user
-
-        add, remove = self.get_institutions_to_add_remove(
-            institutions=instance['data'],
-            new_institutions=validated_data['data']
-        )
-
-        for inst in remove:
-            if not user.is_affiliated_with_institution(inst) and not node.has_permission(user, 'admin'):
```
Comment (brianjgeiger) [inline api/nodes/serializers.py]: And here

File: api/nodes/serializers.py (hunk 1 above)
Comment (brianjgeiger) [reply api/nodes/serializers.py]: Is that helpful/better? I need the inst below unless I refactor.

File: api/nodes/serializers.py (hunk 2 above)
Comment (brianjgeiger) [reply api/nodes/serializers.py]: On the node view's permissions class? This isn't going through the node-institutions view, so I'm not sure that's correct in this case.

**Pull request merged**